torch/csrc/jit/c10_ops/layer_norm.cpp

```diff
@@ -29,35 +29,25 @@ RegisterOperators reg({
   Operator(
     "caffe2::layer_norm_dont_use_this_op_yet(Tensor input, int axis, float epsilon) -> (Tensor, Tensor, Tensor)",
     [](Stack& stack) {
-        ArrayRef<IValue> inputs = last(stack, 3);
-        Tensor input = std::move(inputs[0]).toTensor();
-        const IValue& axis = inputs[1];
-        const IValue& epsilon = inputs[2];
-        drop(stack, 3);
-
-        if (input.requires_grad()) {
+        Tensor tensor_input = std::move(stack[stack.size()-3]).toTensor();
+        if (tensor_input.requires_grad()) {
           throw std::runtime_error("Autograd not yet supported for c10 ops.");
         }
+        auto device = tensor_input.device();
+        torch::jit::peek(stack, 0, 3) = torch::autograd::Variable(std::move(tensor_input)).data();
+
+        // push output fields as outputs to stack
+        push(stack, at::empty({0}, device), at::empty({0}, device), at::empty({0}, device));
```

Comment:
sorry, I'm really confused on what your desired contract for Stack is. If you're saying that we want to do Stack, then it makes sense for the kernel to push the outputs to it (after popping inputs). Current movement of outputs is really weird.

Reply:
Yes, this is currently abusing the stack. I don't really have a better solution (yet). Assuming a JIT schema like this:

    operator(Tensor in1, Tensor in2) -> Tensor

the usual contract is that the kernel pops both tensors from the stack, and then pushes one tensor for the result.

In caffe2, this doesn't work because it needs preallocated output tensors as inputs. To fix that, we could do this:

  operator(Tensor in1, Tensor in2, Tensor out) -> Null

but then, the call side would have no way of reading the outputs written to by the kernel. We actually need to still have the outputs on the stack when returning, so my current plan is to push all inputs *and* outputs to the stack when calling a caffe2 kernel, and also leave them all there on the stack. This is equivalent to an operator schema of

    operator(Tensor in1, Tensor in2, Tensor out) -> (Tensor, Tensor, Tensor)

The frontend user calling into this op, however, doesn't want to deal with that. They want a plain

    operator(Tensor in1, Tensor in2) -> Tensor

so I need to wrap between these two kernels, which is done by moving the outputs down the stack.

It's admittedly weird and the last word here is not spoken yet. I'm currently considering having the caffe2 kernel do:

    operator(Tensor in1, Tensor in2, TensorList? outputs) -> Tensor

which would take an optional list of output tensors as input and could align what the user expects to see and what the kernel implements. The downside of this, however, is that it needs an additional allocation for the TensorList. Having it not be a tensor list but multiple optional output tensor arguments gets weird when some of them are set and some are not, which the kernel can't necessarily handle. I'm still trying to find a better way here.

Reply:
Can all of that be handled in operator invocation scaffolding for C2 ops? The rationale is that current JIT invocation interface (Stack*) indeed doesn't support persistent output tensors. If it did - we'd likely need to change that interface. From this perspective, can't we just do something like following (assuming the schema is the regular one):

```
void invoke_with_stack(Stack*) {
caffe2_op->SetInputs(peekSlice(stack, num_inputs));
caffe2_op->outputs.resize(empty blobs..); // reset outputs to uninitialized state or something like it (or can be raw ivalue)
caffe2_op->run(); // run would execute and initialize the output tensors from scratch
stack->pop(num_inputs);
stack->push_move(std::move(caffe2_op->outputs)); // just move those outputs back to stack
}
```